tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -1130,6 +1211,31 @@ private[optimizer] abstract class OptimizerCore(
         BinaryOp(op, finishTransformExpr(lhs), finishTransformExpr(rhs))
       case PreTransLocalDef(localDef) =>
         localDef.newReplacement
+
+      /* In general, it is not OK to allocate a new instance of an inlined
+       * class from its record value, because that can break object identity
+       * (not to mention we have no idea what the primary constructor does).
+       * However, for RuntimeLong specifically, it is OK. It is useful to do
+       * so because it allows us not to cancel the original stack allocation
+       * of the Long value, which means that all previous usages of it can
+       * stay on stack.
+       *
+       * We do something similar in LocalDef.newReplacement.
+       */
+      case PreTransRecordTree(tree, tpe, _)
+          if tpe.base == ClassType(LongImpl.RuntimeLongClass) =>
+        tree match {
+          case RecordValue(_, List(lo, hi)) =>
+            createNewLong(lo, hi)
+          case recordVarRef: VarRef =>
+            createNewLong(recordVarRef)
+          case _ =>
+            val varRefIdent = Ident(freshLocalName("x", mutable = false))
+            val recordVarDef =
+              VarDef(varRefIdent, tree.tpe, mutable = false, tree)
+            Block(recordVarDef, createNewLong(recordVarDef.ref))
+        }
+
```

Comment:
Yes. Well that exact snippet was already "bad", because `1L` is a constant and so `l` is constant-folded anyway, causing three allocations. 

But now this is true in more cases, for example if we replace `val l = x + 1L`.